assembly/src/parsers/io_ops.rs

```diff
@@ -90,14 +113,62 @@ pub fn parse_env(span_ops: &mut Vec<Operation>, op: &Token) -> Result<(), Assemb
 // NON-DETERMINISTIC INPUTS
 // ================================================================================================
 
-/// TODO: implement
-pub fn parse_read(_span_ops: &mut Vec<Operation>, _op: &Token) -> Result<(), AssemblyError> {
-    unimplemented!()
+/// Appends the number of READ operations specified by the operation's immediate value to the span
+/// block. This pushes the specified number of items from the advice tape onto the stack.
+///
+/// # Errors
+///
+/// Returns an AssemblyError if the instruction is invalid, malformed, missing the required
+/// parameter, or does not match the expected operation. Returns an invalid_param AssemblyError if
+/// the parameter is not a decimal value.
+pub fn parse_read(span_ops: &mut Vec<Operation>, op: &Token) -> Result<(), AssemblyError> {
+    // validate the operation
+    validate_op_len(op, 2, 1, 1)?;
+    let op_parts = op.parts();
+    if op_parts[0] != "adv" || op_parts[1] != "push" {
+        return Err(AssemblyError::unexpected_token(op, "adv.push.{param}"));
+    }
+
+    // parse the number of items to get from advice tape as a decimal
+    let param_idx = 2;
+    let param_value = op.parts()[param_idx];
+    let n = match param_value.parse::<u64>() {
+        Ok(i) => i,
+        Err(_) => return Err(AssemblyError::invalid_param(op, param_idx)),
+    };
+
+    // read n items from the advice tape and push then onto the stack
+    for _ in 0..n {
+        span_ops.push(Operation::Read);
+    }
+
+    Ok(())
 }
 
-/// TODO: implement
-pub fn parse_readw(_span_ops: &mut Vec<Operation>, _op: &Token) -> Result<(), AssemblyError> {
-    unimplemented!()
+/// Appends 4 PAD operations to the span block to make space for elements from the advice tape,
+/// then appends a READW operation to read a word from the advice tape and overwrite the top 4
+/// elements of the stack.
+///
+/// # Errors
+///
+/// Returns an AssemblyError if the instruction is invalid, malformed, or doesn't match the
+/// expected operation.
+pub fn parse_readw(span_ops: &mut Vec<Operation>, op: &Token) -> Result<(), AssemblyError> {
+    validate_op_len(op, 2, 0, 0)?;
+    let op_parts = op.parts();
+    if op_parts[0] != "adv" || op_parts[1] != "loadw" {
+        return Err(AssemblyError::unexpected_token(op, "adv.loadw"));
+    }
+
+    // Make space for 4 elements on the stack
+    for _ in 0..4 {
+        span_ops.push(Operation::Pad);
+    }
+
+    // load a word from the advice tape
+    span_ops.push(Operation::ReadW);
+
+    Ok(())
 }
```

Comment:
This will probably need to be merged with the function above. The parser will be triggered when the first part of an instruction is `adv` and then we can figure out whether we are doing a `push` or a `load`.